### PR TITLE
[Bugfix] Spy on unnamed events too

### LINKED ISSUE #717: `spy_events` not working on events emitted with `emit_event_syscall`
failing test example https://github.com/foundry-rs/starknet-foundry/commit/1d2ed5428f337f0e02cf1fd148dd8514b0e18c37

### PR DESCRIPTION
**Pull request by war-in**
Closes #717 

## Introduced changes

- events emitted with `emit_event_syscall` can be unnamed
- spy_events handles this now

## Breaking changes

- Introduce `NamedEvent` and `UnnamedEvent` structs

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: snforge_std/src/cheatcodes/events.cairo [hunk 1]
```diff
@@ -102,10 +151,16 @@ fn assert_if_emitted(ref self: EventSpy, event: Event) -> bool {
...
-            && event.from == *emitted_events.at(j).from
-            && @event.keys == emitted_events.at(j).keys
-            && @event.data == emitted_events.at(j).data {
+        let emitted = match event.clone() {
+            Event::Named(named_event) => {
+                let mut cloned_event = named_event.clone();
+                cloned_event.name = event_name_hash(cloned_event.name);
+                Event::Named(cloned_event) == emitted_events.at(j).clone()
+            },
+            Event::Unnamed(unnamed_event) => Event::Unnamed(unnamed_event) == emitted_events.at(j).clone()
```
Comment (piotmag769) [inline snforge_std/src/cheatcodes/events.cairo:160]:
```suggestion
           unnamed => unnamed == emitted_events.at(j).clone()
```

File: crates/forge/tests/data/contracts/spy_events_checker.cairo [hunk 2]
```diff
@@ -74,5 +76,9 @@ mod SpyEventsChecker {
             self.emit(Event::SecondEvent(SecondEvent { some_data, some_more_data }));
             self.emit(Event::ThirdEvent(ThirdEvent { some_data, some_more_data, even_more_data }));
         }
+
+        fn emit_event_syscall(ref self: ContractState, some_data: felt252) {
+            starknet::emit_event_syscall(array![].span(), array![some_data].span()).unwrap_syscall();
+        }
```
Comment (piotmag769) [inline crates/forge/tests/data/contracts/spy_events_checker.cairo:82]: We could emit some keys to check if we handle them correctly too

**Review by piotmag769 — COMMENTED** [review]
I believe we should also update docs

File: crates/forge/tests/data/contracts/spy_events_checker.cairo (hunk 2 above)
Comment (war-in) [reply crates/forge/tests/data/contracts/spy_events_checker.cairo:82]: The first key will be interpreted as a hashed name. There is no other option to handle named events

File: snforge_std/src/cheatcodes/events.cairo [hunk 3]
```diff
@@ -28,14 +28,58 @@ fn event_name_hash(name: felt252) -> felt252 {
...
+        match self {
+            Event::Named(NamedEvent { name, .. }) => Option::Some(name),
+            Event::Unnamed(UnnamedEvent { from, .. }) => Option::None,
+        }
+    }
+
+    fn keys(self: Event) -> Array<felt252> {
+        match self {
+            Event::Named(NamedEvent { keys, .. }) => keys,
+            Event::Unnamed(UnnamedEvent { keys, .. }) => keys,
```
[the remark this answers is not in the document]
Comment (war-in) [reply snforge_std/src/cheatcodes/events.cairo:56]: Unfortunately not :(

File: snforge_std/src/cheatcodes/events.cairo (hunk 1 above)
Comment (war-in) [reply snforge_std/src/cheatcodes/events.cairo:160]: We have to match `Event::Unnamed` and then compare it. `unnamed` is not a variant

File: crates/forge/tests/data/contracts/spy_events_checker.cairo (hunk 2 above)
Comment (Arcticae) [reply crates/forge/tests/data/contracts/spy_events_checker.cairo:82]: Why is there a field in `UnnamedEvent` struct for keys then? That seems counterintuitive. Perhaps we should serialize the events in the same way in cairo (with a uniform `EventData`) and instead serialize the passed one, to the uniform data. We could then serialize the user-passed event into uniform representation, and then compare with ones that were captured. WDYT?

**Pull request closed without merging**